src/Microsoft.AspNet.OData.Shared/Formatter/Deserialization/DeserializationHelpers.cs

```diff
@@ -10,18 +10,41 @@
 using System.Runtime.Serialization;
 using Microsoft.AspNet.OData.Builder;
 using Microsoft.AspNet.OData.Common;
+using Microsoft.AspNet.OData.Extensions;
 using Microsoft.OData;
 using Microsoft.OData.Edm;
+using System.Linq;
 
 namespace Microsoft.AspNet.OData.Formatter.Deserialization
 {
     internal static class DeserializationHelpers
     {
         internal static void ApplyProperty(ODataProperty property, IEdmStructuredTypeReference resourceType, object resource,
             ODataDeserializerProvider deserializerProvider, ODataDeserializerContext readContext)
-        { 
-            IEdmProperty edmProperty = resourceType.FindProperty(property.Name);
-                       
+        {
+            IEdmProperty edmProperty;
+
+            // Extract edmProperty using a Case-Sensitive match
+            edmProperty = resourceType.FindProperty(property.Name);
+
+            if (readContext != null && readContext.EnableCaseInsensitiveModelBinding && edmProperty == null)
+            {
+                // if edmProperty is null, we try a Case-Insensitive match
+                var sProperties = resourceType.StructuralProperties().Where(k => k.Name.Equals(property.Name, StringComparison.OrdinalIgnoreCase));
+
+                // We throw an exception when we have more than 1 case-insensitive matches and no case sensitive matches.
+                if (sProperties.Count() > 1)
+                {
+                    throw new ODataException(Error.Format(SRResources.CannotDeserializeUnknownProperty, property.Name, resourceType.Definition));
+                }
+
+                if (sProperties.Count() == 1)
+                {
+                    var sProperty = sProperties.Single();
+                    edmProperty = resourceType.FindProperty(sProperty.Name);
+                }
```

Comment:
calling count twice against an IQueryable is expensive -- it actually evaluates the expensive LINQ query twice -- then a third time once you do the .Single(). Better would be to loop through the structural properties doing the case insensitive match -- if you find one keep going, if you find a second then error.